pkg/kubelet/kubelet.go

```diff
@@ -3354,65 +3352,54 @@ func (kl *Kubelet) convertStatusToAPIStatus(pod *api.Pod, podStatus *kubecontain
+		//   * If the container is Running, the start error will be dismissed.
+		//   * If the container is Terminated, the container will be set as *Waiting* with start error, original Terminated
+		//   state will become LastTerminationState.
+		// However, we only promised best-effort Waiting state, so these are acceptable.
+		if !kubecontainer.ShouldContainerBeRestarted(&container, pod, podStatus) {
 			continue
 		}
-		var containerStatus api.ContainerStatus
-		containerStatus.Name = container.Name
-		containerStatus.Image = container.Image
-		if oldStatus, found := oldStatuses[container.Name]; found {
-			// Some states may be lost due to GC; apply the last observed
-			// values if possible.
-			containerStatus.RestartCount = oldStatus.RestartCount
-			containerStatus.LastTerminationState = oldStatus.LastTerminationState
+		containerStatus, found := statuses[container.Name]
+		if found {
+			_, _, ok := kl.reasonCache.Get(uid, container.Name)
+			if !ok {
+				// In fact, we could also apply Waiting state here, but it is less informative than Terminated,
+				// and the container will be Running soon, so we prefer Terminated here.
+				continue
+			}
+			containerStatus.LastTerminationState = containerStatus.State
+		} else {
+			containerStatus = &api.ContainerStatus{Name: container.Name, Image: container.Image}
+			if oldStatus, found := oldStatuses[container.Name]; found {
+				// Some states may be lost due to GC; apply the last observed values if possible.
+				containerStatus.RestartCount = oldStatus.RestartCount
+				containerStatus.LastTerminationState = oldStatus.LastTerminationState
+			}
 		}
-		reason, _, ok := kl.reasonCache.Get(uid, container.Name)
-
+		reason, message, ok := kl.reasonCache.Get(uid, container.Name)
 		if !ok {
 			// default position for a container
 			// At this point there are no active or dead containers, the reasonCache is empty (no entry or the entry has expired)
 			// its reasonable to say the container is being created till a more accurate reason is logged
-			containerStatus.State = api.ContainerState{
-				Waiting: &api.ContainerStateWaiting{
-					Reason:  fmt.Sprintf("ContainerCreating"),
-					Message: fmt.Sprintf("Image: %s is ready, container is creating", container.Image),
-				},
-			}
-		} else if reason == kubecontainer.ErrImagePullBackOff ||
-			reason == kubecontainer.ErrImageInspect ||
-			reason == kubecontainer.ErrImagePull ||
-			reason == kubecontainer.ErrImageNeverPull ||
-			reason == kubecontainer.RegistryUnavailable {
-			// mark it as waiting, reason will be filled bellow
-			containerStatus.State = api.ContainerState{Waiting: &api.ContainerStateWaiting{}}
-		} else if reason == kubecontainer.ErrRunContainer {
-			// mark it as waiting, reason will be filled bellow
-			containerStatus.State = api.ContainerState{Waiting: &api.ContainerStateWaiting{}}
+			reason = fmt.Errorf("ContainerCreating")
+			message = fmt.Sprintf("Image: %s is ready, container is creating", container.Image)
```

Comment:
Do we know for sure that the image is ready here?  This could occur before image pull, no?